kits/arms/experimental/HebiArm.m

```diff
@@ -0,0 +1,292 @@
+                   error(['Expected Aux: ' num2str(length(this.auxTime)-1) 'xM matrix']);
+               end
+            end
+            
+        end
+        
+        function value = isAtGoal(this)
+            value = this.getProgress() >= 1;
+        end
+        
+        function [progress] = getProgress(this)
+            % returns normalized [0-1] time of where trajectory is
+            % currently at
+            if isempty(this.traj) || isempty(this.state)
+                error('trajectory not initialized');
+            end
+            
+            t = this.state.time-this.trajStartTime;
+            progress = min(t / this.traj.getDuration(),1);
+            
+        end
+        
+        function [] = update(this)
+            % reads feedback and updates internal state
+            
+            % Read feedback
+            newState = struct();
+            newState.fbk = this.group.getNextFeedbackFull();
+            newState.time = newState.fbk.time;
+            newState.numDoF = this.kin.getNumDoF();
+            
+            % Always compensate for accelerations due to gravity
+            newState.cmdEffort = this.getGravCompEfforts(newState.fbk);
+            
+            % Add pos/vel/accel commands
+            newState.cmdAux = [];
+            if ~isempty(this.traj)
+                
+                % Evaluate trajectory state
+                t = min(newState.time - this.trajStartTime, this.traj.getDuration());
+                [ newState.cmdPos,  newState.cmdVel,  newState.cmdAccel] = this.traj.getState(t);
+                
+                % Compensate for joint accelerations
+                newState.cmdEffort = this.kin.getDynamicCompEfforts(...
+                    newState.fbk.position, ...
+                    newState.cmdPos, ...
+                    newState.cmdVel, ...
+                    newState.cmdAccel) + newState.cmdEffort;
+                
+                % Add aux (e.g. gripper state)
+                if ~isempty(this.aux)
+                    auxRow = find(t >= this.auxTime, 1, 'last' );
+                    newState.cmdAux = this.aux(auxRow, :);
+                end
+                
+                newState.trajTime = t;
+                
+            else
+                newState.cmdPos = [];
+                newState.cmdVel = [];
+                newState.cmdAccel = [];
+                newState.trajTime = [];
+            end
+            
+            % Call plugins (FK, Jacobians, End-Effector XYZ, etc.)
+            this.state = newState;
+            for i=1:length(this.plugins)
+                this.plugins{i}.update(this);
+            end            
+            
+            % TODO: --> user code
+            % * add effort offsets for e.g. spring.
+            % * add efforts to add an end effector wrench (local or world coordinates?)
+            % * add impedance controller
+            % * remove efforts if all zeros? e.g. strategy 3
```

Comment:
The impedance controller and effort offsets are already done, right?  Should we take out these TODOs?

Reply:
an oversight. thanks